app/app.rb

```diff
@@ -18,10 +20,33 @@ class MakersBnb < Sinatra::Base
                 email: params[:email],
                 password: params[:password],
                 password_confirmation: params[:password_confirmation])
+    if user.saved? 
     session[:user] = user.id
     redirect('/spaces')
+    else 
+    flash.next[:errors] = user.errors.full_messages
+    redirect('/')
+    end
   end
 
+  get '/user/sign_in' do 
+    erb :sign_in
+  end
+
+  post '/user/sign_in' do 
+    user = User.authenticate(params[:email],params[:password])
+    if user == -1
```

Comment:
Went with a more nuanced if-else tree in order to address different errors on sign-in (non-existent user vs. incorrect password).